src/libraries/System.Private.CoreLib/src/System/Number.DecimalIeee754.cs

```diff
@@ -0,0 +1,255 @@
+            where TSignificand : IBinaryInteger<TSignificand>
+        {
+            public bool Signed { get; }
+            public int Exponent { get; }
+            public TSignificand Significand { get; }
+
+            public DecimalIeee754(bool signed, int exponent, TSignificand significand)
+            {
+                Signed = signed;
+                Exponent = exponent;
+                Significand = significand;
+            }
+        }
+
+        internal static DecimalIeee754<TSignificand> UnpackDecimalIeee754<TDecimal, TSignificand, TValue>(TValue value)
+            where TDecimal : unmanaged, IDecimalIeee754UnpackInfo<TDecimal, TSignificand, TValue>
+            where TSignificand : IBinaryInteger<TSignificand>
+            where TValue : IBinaryInteger<TValue>
+        {
+            bool signed = (value & TDecimal.SignMask) != TValue.Zero;
+            TSignificand significand;
+            int exponent;
+
+            if ((value & TDecimal.G0G1Mask) == TDecimal.G0G1Mask)
+            {
+                exponent = TDecimal.ConvertToExponent((value & TDecimal.G2ToGwPlus3ExponentMask) >> (TDecimal.NumberBitsSignificand + 1));
+                significand = TDecimal.ConvertToSignificand((value & TDecimal.GwPlus4SignificandMask) | TDecimal.MostSignificantBitOfSignificandMask);
+            }
+            else
+            {
+                exponent = TDecimal.ConvertToExponent((value & TDecimal.G0ToGwPlus1ExponentMask) >> (TDecimal.NumberBitsSignificand + 3));
+                significand = TDecimal.ConvertToSignificand(value & TDecimal.GwPlus2ToGwPlus4SignificandMask);
+            }
+
+            return new DecimalIeee754<TSignificand>(signed, exponent - TDecimal.Bias, significand);
+        }
+
+        internal static int CompareDecimalIeee754<TDecimal, TSignificand, TValue>(TValue currentValue, TValue otherValue)
+            where TDecimal : unmanaged, IDecimalIeee754UnpackInfo<TDecimal, TSignificand, TValue>
+            where TSignificand : IBinaryInteger<TSignificand>
+            where TValue : IBinaryInteger<TValue>
+        {
+            if (currentValue == otherValue)
+            {
+                return 0;
+            }
+            DecimalIeee754<TSignificand> current = UnpackDecimalIeee754<TDecimal, TSignificand, TValue>(currentValue);
+            DecimalIeee754<TSignificand> other = UnpackDecimalIeee754<TDecimal, TSignificand, TValue>(otherValue);
+
+            if (current.Signed && !other.Signed) return -1;
+
+            if (!current.Signed && other.Signed) return 1;
+
+            if (current.Exponent > other.Exponent)
+            {
+                return current.Signed ? -InternalUnsignedCompare(current, other) : InternalUnsignedCompare(current, other);
+            }
+
+            if (current.Exponent < other.Exponent)
+            {
+                return current.Signed ? InternalUnsignedCompare(other, current) : -InternalUnsignedCompare(current, other);
+            }
```

Comment:
You have `current, other` for 3 of them and `other, current` for the fourth; this looks like a bug.

I would expect you're either doing `Cmp(x, y)` vs `Cmp(y, x)` -or- you're doing `Cmp(x, y)` and `-Cmp(x, y)`

I'd try and simplify this down to one call instead, with a comment elaborating why you're doing the swap, such as:
```csharp
if (current.Signed)
{
    // Both values are negative and so we need to swap them to ensure
    // the comparison gives the right result. This is because we're doing
    // an unsigned comparison and so for `Cmp(-2, -3)` we'd get `Cmp(2, 3)`
    // and the result would be `-1`. But since it's negative we really want
    // `+1` since `-2 > -3`.
    (current, other) = (other, current);
}
return Cmp(current, other);
```